Support StartTLS mail servers in IMAP sensors

amelchio opened the pull request:
## Description:

Being able to disable SSL lets the imap sensor work with servers expecting STARTTLS.

If this is acceptable I will update the documentation and the imap_email_content sensor as well.

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: imap
    server: !secret mail_server
    port: 143
    username: !secret mail_username
    password: !secret mail_password
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

amelchio committed:
Add SSL config option for IMAP sensor

An IMAP server on port 143 expects the STARTTLS command, not a SSL
wrapping of the entire connection.

So disable SSL except for the official "IMAP over SSL" on port 993.

Port 993 remains the default, keeping compatibility.

pvizeli commented on homeassistant/components/sensor/imap.py:
```diff
@@ -27,6 +27,7 @@
     vol.Required(CONF_PASSWORD): cv.string,
     vol.Required(CONF_SERVER): cv.string,
     vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_SSL): cv.boolean,
```
Make a `DEFAULT_SSL = True` and set this for default.

pvizeli reviewed (changes requested):
Make a `DEFAULT_SSL = True` and set this for default.

amelchio commented:
Hmm, the SSL default depends on the selected port, only being True for "IMAP over SSL" (port 993). Setting an explicit default breaks that.
```
self._ssl = (port == 993) if ssl is None else ssl
```
Is that a bad way of doing it?

pvizeli commented:
Yes. On HomeAssistant is security opt-out and not opt-in. So we need explicit deactivate SSL and default is the hight possible security options.

amelchio commented:
Right, I believed that the plaintext connection would auto-upgrade with StartTLS, but I now see that I must have been confused. I will rework this and come back with a different approach. Thanks for reviewing!

amelchio committed:
Support IMAP connections on the standard port

The modern way to use IMAP is to connect on the classic port 143 and
upgrade the connection with the STARTTLS command.

This commit adds support for such a server while keeping the default
unchanged at port 993 (IMAP over SSL).

There is still no way to opt out of the encryption.

amelchio committed:
Revert "Add SSL config option for IMAP sensor"

amelchio commented:
Now I reverted the entire commit and made a new one that insists on encryption. Thus, no new config value is needed.

I am unsure how to present this revised work so let me know if I should squash the commits, create a new PR or whatever.

pvizeli commented:
Please add unittest for it to check if it will work

pvizeli reviewed (changes requested):
That make it inpossible to use a other port for ssl. Make it as simple option to choise if you want use `starttls`

pvizeli commented:
https://home-assistant.io/components/notify.smtp/

pvizeli commented:
Do you want address comments?

pvizeli commented:
Please reopen if you want finish it

amelchio commented:
Yeah, I lost the motivation to complete this. I disagree with an implementation that does not guess the protocol based on the port number since it will almost always guess right.

Also, ```notify.smtp``` is insecure by default so that is a bad example to follow.

I wanted to propose a ```security``` option with values:
- ```auto``` (default)
- ```ssl```
- ```starttls```
- ```none``` (I think that this one should not be supported)

But then you told me to write test cases for an untested area in an application that is unfamiliar to me and it ended up too much trouble to get a five line patch accepted.
